fix thread number parameter

hach1yon opened the pull request:
… thread number parameter to Windows Eventlog Parser Library

## What Changed

- remove thread number parameter from Computer Metrics Command
- apply thread number parameter to Windows Event Log Parser Library

## Evidence

```
C:\Users\takai\dev\hayabusa>cargo run computer-metrics
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.20s
     Running `target\debug\hayabusa.exe computer-metrics`
error: the following required arguments were not provided:
  <--directory <DIR>|--file <FILE>|--live-analysis>

Usage: hayabusa.exe computer-metrics <--directory <DIR>|--file <FILE>|--live-analysis>

For more information, try '--help'.
error: process didn't exit successfully: `target\debug\hayabusa.exe computer-metrics` (exit code: 2)
```

Commit:
remove thread number parameter from Computer Metics Command and apply thread number parameter to Windows Eventlog Parser Library

fukusuket reviewed (approved):
LGTM!!🚀

fukusuket commented:
@hach1yon @YamatoSecurity 
Sorry! I overlooked this in the review😅
It seems that we also need to modify the following code to pass the `-t` parameter to the `search` command — 
what do you think?
- https://github.com/Yamato-Security/hayabusa/blob/hajime/fix_multithread_parameter/src/detections/configs.rs#L785
- https://github.com/Yamato-Security/hayabusa/blob/hajime/fix_multithread_parameter/src/detections/configs.rs#L922-L933

hach1yon commented:
I'll check it.

fukusuket commented:
@hach1yon 
Thank you for checking! 
I also confirmed with print-debug that the `csv-timeline`  works as expected!

for the `search`, the value printed is `0` instead of the Thread number specified with `-t`.
```
% ./hayabusa search -d ../data/windows/hayabusa-sample-evtx-main/ -o searc.csv -k "mimikatz" -q -C -t 10
...
[00:00:00] 18 / 598 ⠁ [=>                                      ] 3%

"../data/windows/hayabusa-sample-evtx-main/DeepBlueCLI/powersploit-security.evtx" (68.0 KiB)                                                              
Thread num:0
```

This issue can be fixed as follows.
```rust
/// config情報からthread_numberの情報を抽出する関数
fn check_thread_number(config: &Config) -> Option<usize> {
    match config.action.as_ref()? {
        Action::CsvTimeline(opt) => opt.output_options.detect_common_options.thread_number,
        Action::JsonTimeline(opt) => opt.output_options.detect_common_options.thread_number,
        Action::LogonSummary(opt) => opt.detect_common_options.thread_number,
        Action::EidMetrics(opt) => opt.detect_common_options.thread_number,
        Action::ExtractBase64(opt) => opt.detect_common_options.thread_number,
        Action::PivotKeywordsList(opt) => opt.detect_common_options.thread_number,
        Action::LogMetrics(opt) => opt.detect_common_options.thread_number,
        Action::Search(opt) => opt.thread_number,  # add this line
        _ => None,
    }
}
```

As shown below, the Thread number specified with `-t` was printed correctly in the `search` command.
```
% ./hayabusa search -d ../data/windows/hayabusa-sample-evtx-main/ -o searc.csv -k "mimikatz" -q -C -t 10
...

"../data/windows/hayabusa-sample-evtx-main/DeepBlueCLI/powersploit-security.evtx" (68.0 KiB)                                                              
Thread num:10
```
>  I also executed the command cargo run --release -- search --file C:\Users\takai\dev\hayabusa-sample-evtx\DeepBlueCLI\many-events-security.evtx -t 1 --keyword hajime, but Rayon still seemed to create 24 threads instead of 1.

Probably, with this fix, 24 threads will also be executed in the case of the `search` command. 🤔

fukusuket commented:
@hach1yon @YamatoSecurity 
Once the issue where the number of threads specified with `-t` in the `search` command is not applied is fixed, I think this PR can be merged!

hach1yon commented:
> I checked the code in the `evtx` crate related to thread specification, and found that `thread_num` is used to divide the evtx records chunks as shown below.
> 
> * https://github.com/omerbenamram/evtx/blob/db408b18e13454c0fdfe359ed23220000ca0f798/src/evtx_parser.rs#L411-L463
> 
> Since it doesn’t directly create threads, it may internally create up to the maximum number of threads based on the number of CPU cores...? 🤔 (The Rayon documentation mentioned that it uses the number of logical CPU cores)
> 
> * https://docs.rs/rayon/latest/rayon/struct.ThreadPoolBuilder.html#method.num_threads

Thanks. As far as I know, this code should work with multiple threads. I'll go over why rayon doesn't work in parallel.

Commit:
fix bug: thread number parameter for search command wasn't passed to EventParser Library.

hach1yon commented:
@YamatoSecurity @fukusuket 
I fixed a search command parameter issue.

As we discussed, the number of thread in Windows EventParser Library cannot be controlled although I fixed a search command parameter issue. The problem occurs if you use the other commands. I'll take care of it in another PR.

fukusuket commented:
@hach1yon Thank you for fixing this! LGTM!!🚀

YamatoSecurity committed:
update changelog

YamatoSecurity commented:
@fukusuket Sorry there is a conflict from merging your PR for the checksum validation. Could you take a look at it?

fukusuket committed:
Merge branch 'main' into hajime/fix_multithread_parameter

YamatoSecurity reviewed (approved):
@hach1yon LGTM! Thanks so much for fixing this!!
@fukusuket Thanks so much for checking and fixing the merge conflict!
